File: education/pascal.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Pascal
{
public:
  auto value_of(int x, int y) -> int;
  auto values_of(int row)     -> vector<int>;
  auto line_of(int row)       -> string;
  auto lines_upto(int row)    -> string;
};

auto Pascal::value_of(int x, int y) -> int
{
  if (x < 1 || y < 1 || y > x)
    return 0;

  if (x == 1)
    return 1;

  return value_of(x-1, y-1) + value_of(x-1, y);
}

auto Pascal::values_of(int row) -> vector<int>
{
  vector<int> v;

  if (row > 0)
    for (int col=1; col<=row; col++)
      v.push_back(value_of(row, col));

  return v;
}

auto Pascal::line_of(int row) -> string
{
  string res;
  for (int i=1; i<=row; ++i)
  {
    char buf[1024] = { 0 };
    sprintf (buf, "%6d", value_of(row, i));
    res += string(buf);
  }

  return res;
}

auto Pascal::lines_upto(int row) -> string
{
  string res;
  for (int i=1; i<=row; i++)
    res += string(3*(row-i), ' ') + line_of(i) + "\n";

  return res;
}

int main(int argc, const char *argv[])
{
  Pascal p;

  cout << p.lines_upto(7);

  return 0;
}

File: education/pascal.rb
#!/usr/bin/env ruby

require 'test/unit'

class Pascal
  def value_of(x, y)
    return 0 if x < 1 or y < 1 or y > x
    return 1 if x == 1 
    return value_of(x-1, y-1) + value_of(x-1, y)
  end
  
  def values_of(row)
    return [] if row == 0
    res = []
    1.upto(row) { |col| res << value_of(row, col) }
    res
  end
  
  def line_of(row)
    res = ""; 1.upto(row) { |col| res += sprintf("%6d", value_of(row, col)) }
    res
  end
  
  def lines_upto(row)
    res = ""; 1.upto(row) { |r| res += " " * (row-r)*3 + line_of(r) + "\n" }
    res
  end

end

class TestPascal < Test::Unit::TestCase
  def setup
    @p = Pascal.new
  end

  def test_value_of
    assert_equal(1, @p.value_of(1, 1))
    assert_equal(0, @p.value_of(1, 2))
    assert_equal(1, @p.value_of(2, 1))
    assert_equal(1, @p.value_of(2, 2))
    assert_equal(2, @p.value_of(3, 2))
    assert_equal(1, @p.value_of(3, 3))
    assert_equal(3, @p.value_of(4, 2))
    assert_equal(3, @p.value_of(4, 3))
    assert_equal(10, @p.value_of(6, 3))
  end

  def test_values_of
    assert_equal([], @p.values_of(0))
    assert_equal([1], @p.values_of(1))
    assert_equal([1,1], @p.values_of(2))
    assert_equal([1, 2, 1], @p.values_of(3))
    assert_equal([1, 3, 3, 1], @p.values_of(4))
    assert_equal([1, 4, 6, 4, 1], @p.values_of(5))
  end
  
  def test_values_upto
    puts ""
    puts @p.lines_upto(10)
  end
end
